templated-this.md

```diff
@@ -1,151 +1,315 @@
+
+A common task in writing a class is providing a "getter function" to access a
+contained member. Authors that care about performance want to provide getter
+functions that take advantage of move semantics in the case where the class
+instance is an rvalue, and so end up writing several of these functions, which
+only differ in the cv-ref qualifiers .
+
+
+### There is no way to write a reference qualifier for a lambda expression.
+
+When writing a lambda expression, it is impossible to know whether it is safe to
+move from captured-by-copy members of the closure object. This impacts the
+performance of lambda expressions. One important scenario is when using lambda
+expressions as parts of a chain of asynchronous continuations that carry a value
+through a computation. Without knowing the value category of a closure object
+from inside its function call operator, the implementer has a dilemma: incur a
+copy, or use a move, and impose a requirement that is unenforceable in code and
+sets hidden traps for the unwary.
+
+
+Design Decisions
+----------------
+   
+In addition to solving the existing problems, desirable properties of a solution
+are:
+
+* It should work the same way for member functions and lambda expressions.
+
+* It should work like existing practice as much as possible, while adding as
+  little extra syntax as possible. Using `auto&& this` follows existing patterns
+  for dealing with forwarding references. Optionally adding "this" as the first
+  parameter fits with many programmers' mental model of the `this` pointer being
+  the first parameter to member functions "under the hood" and is comparable to
+  usage in other languages, e.g. Python.
+
+* It should avoid adding extraneous syntax to function declarations. In a world
+  where every function is fast becoming tagged with attributes, constexpr,
+  noexcept specifications, and (this proposal notwithstanding) reference
+  qualifiers, readable, noise-free function declarations are to be prized.
+
+TBD: more here.
+
+
+Proposed Solution
+-----------------
+
+We propose the ability to add an optional first parameter to any member function
+of a class `T`, taking the form `T [const] [volatile] [&|&&] this`.
+
+To facilitate use in generic lambda expressions, this may also be formulated as
+`auto [const] [volatile] [&|&&] this`.
+
+In all cases, the value-category of `this` inside the member function is exactly
+what the existing parameter rules already imply.
+
+This, in the cases where the type of `this` is not a template, maps to the
+existing practice of putting cv-ref qualifiers (that really apply to `*this` in
+current parlance) on the function itself, by writing them behind the parameter
+list, as far as the function signature is concerned.
+
+```cpp
+// Within a class
+
+struct Person {
+  std::string name;
+  
+  // Getter:
+  template <typename T>
+  decltype(auto) GetName(T&& this) { 
+      // T can only deduce to Person (mod cvref) because of visibility
```

Comment:
Is it not the case that the && is a forwarding reference here? If T only deduces to Person, this doesn't replace multiplying differently ref-qualified getters?

Reply:
(you missed the 'mod cvref') - modulo reference qualifiers. It means exactly what you think it means. That said, I clearly need to rephrase it because if you didn't understand what I meant, others won't either.

Reply:
:+1: I am a fan of using standardese where necessary, but also plain english where it is accurate.

Reply:
done. Rephrased comment. Please check if it's better.